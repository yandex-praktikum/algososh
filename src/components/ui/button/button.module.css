.button {
  display: flex;
  align-items: center;
  justify-content: center;
  border: none;
  outline: none;
  border-radius: var(--common-border-radius);
  background-color: var(--bg-button-default);
  padding: 18px 24px 19px 24px;
  transition: var(--common-transition);
  max-height: 60px;
  min-width: 130px;
}

.button:hover:enabled {
  background-color: var(--bg-button-hover);
  cursor: pointer;
}

.button:active:enabled {
  background-color: var(--bg-button-active);
  color: var(--text-color-secondary);
  cursor: pointer;
  fill: var(--text-color-secondary);
}

.button:disabled {
  background-color: var(--bg-button-disabled);
}

.big {
  width: 362px;
}

.small {
  width: 175px;
}

.medium {
  width: 205px;
}

.loader:disabled {
  background-color: var(--bg-button-default);
}

.loader_icon {
  animation: rotation 1s linear infinite normal forwards running;
}

@keyframes rotation {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
