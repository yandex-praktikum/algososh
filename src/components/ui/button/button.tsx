import React from "react";
import styles from "./button.module.css";
import loaderIcon from "../../../images/icons/loader.svg";
import { AscendingIcon } from "../icons/ascending-icon";
import { DescendingIcon } from "../icons/descending-icon";
import { Direction } from "../../../types/direction";

interface ButtonProps extends React.HTMLProps<HTMLButtonElement> {
  text?: string;
  type?: "button" | "submit" | "reset";
  sorting?: Direction;
  linkedList?: "small" | "big" | "medium";
  isLoader?: boolean;
  extraClass?: string;
}

export const Button: React.FC<ButtonProps> = ({
  text,
  extraClass = "",
  type = "button",
  isLoader = false,
  sorting,
  linkedList,
  disabled,
  ...rest
}) => {
  const currentIcon =
    sorting === Direction.Ascending ? <AscendingIcon /> : <DescendingIcon />;
  const className = `text text_type_button text_color_primary ${
    styles.button
  } ${linkedList && styles[linkedList]} ${
    isLoader && styles.loader
  } ${extraClass}`;

  return (
    <button
      className={className}
      type={type}
      disabled={isLoader || disabled}
      {...rest}
    >
      {isLoader ? (
        <img className={styles.loader_icon} src={loaderIcon} alt="Загрузка." />
      ) : (
        <>
          {sorting && currentIcon}
          <p className={`text ${sorting && "ml-5"}`}>{text}</p>
        </>
      )}
    </button>
  );
};
