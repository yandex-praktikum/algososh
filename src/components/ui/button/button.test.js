import React from 'react'
import renderer from 'react-test-renderer'
import { render, screen, fireEvent } from '@testing-library/react'

import { Button } from './button'

it('Кнопка с текстом рендерится без ошибок', () => {
  const tree = renderer
    .create(<Button text='Кнопка' />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('Кнопка без текста рендерится без ошибок', () => {
  const tree = renderer
    .create(<Button />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('Заблокированная кнопка рендерится без ошибок', () => {
  const tree = renderer
    .create(<Button disabled={true} text='Кнопка'/>)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('Кнопка с лоадером рендерится без ошибок', () => {
  const tree = renderer
    .create(<Button isLoader={true}/>)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('Обработчик клика по кнопке рендерится без ошибок', () => {
  const tree = renderer
    .create(<Button text='Нажми' onClick={() => alert('Click!')}/>)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('Обработчик клика по кнопке вызывает корректный alert', () => {
  // eslint-disable-next-line
  window.alert = jest.fn()

  render(<Button text='Нажми' onClick={() => alert('Click!')}/>)

  fireEvent.click(screen.getByText('Нажми'))

  expect(window.alert).toHaveBeenCalledWith('Click!')
})
