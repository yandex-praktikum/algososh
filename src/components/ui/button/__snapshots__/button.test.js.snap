// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Заблокированная кнопка рендерится без ошибок 1`] = `
<button
  className="text text_type_button text_color_primary button undefined false "
  disabled={true}
  type="button"
>
  <p
    className="text undefined"
  >
    Кнопка
  </p>
</button>
`;

exports[`Кнопка без текста рендерится без ошибок 1`] = `
<button
  className="text text_type_button text_color_primary button undefined false "
  type="button"
>
  <p
    className="text undefined"
  />
</button>
`;

exports[`Кнопка с лоадером рендерится без ошибок 1`] = `
<button
  className="text text_type_button text_color_primary button undefined loader "
  disabled={true}
  type="button"
>
  <img
    alt="Загрузка."
    className="loader_icon"
    src="loader.svg"
  />
</button>
`;

exports[`Кнопка с текстом рендерится без ошибок 1`] = `
<button
  className="text text_type_button text_color_primary button undefined false "
  type="button"
>
  <p
    className="text undefined"
  >
    Кнопка
  </p>
</button>
`;

exports[`Обработчик клика по кнопке рендерится без ошибок 1`] = `
<button
  className="text text_type_button text_color_primary button undefined false "
  onClick={[Function]}
  type="button"
>
  <p
    className="text undefined"
  >
    Нажми
  </p>
</button>
`;
